src/System.Private.ServiceModel/src/System/ServiceModel/Channels/CommunicationObject.cs

```diff
@@ -758,14 +758,22 @@ internal void ThrowPending()
 
         internal protected virtual Task OnCloseAsync(TimeSpan timeout)
         {
-            Contract.Requires(false, "OnCloseAsync needs to be implemented on derived classes");
-            return TaskHelpers.CompletedTask();
+            // Derived types aware of IAsyncCommunicationObject should override OnCloseAsync.
+            // However, because IAsyncCommunicationObject is internal, external implementations
+            // such as custom channels and factories cannot override it, yet still expect their OnClose
+            // logic to be called.  Moreover, we cannot know BeginClose is implemented, so run the
```

Comment:
Reviewers: FYI, I experimentally made this logic call OnBeginOpen and OnBeginClose via Task.Factory.FromAsync, which logically makes more sense.  However we have some communication object's that throw PNSE from their async paths (e.g. SslStreamSecurityUpgradeProvider).  If we believe calling OnBeginOpen and OnBeginClose is the better fix, I will need to augment this PR to fix our own PNSE's.  If there is a chance customer's custom channels also throw from their async paths, the fix here to call the sync OnOpen and OnClose might be the most conservative fix.  Thoughts welcome.